From 88863c59c5fdc0da38bac338d9f164d4a0259f57 Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Mon, 18 Aug 2025 17:17:57 +0200
Subject: [PATCH 01/13] gluon-core: rename device_uses_11a to device_uses_band

this is a more appropriate naming as the function checks if a radio uses a band
it also is more generic to check for other band usage as well
---
 .../luasrc/lib/gluon/config-mode/wizard/0250-outdoor.lua      | 2 +-
 package/gluon-core/luasrc/lib/gluon/upgrade/180-outdoors      | 2 +-
 package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua      | 4 ++--
 .../luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua  | 2 +-
 4 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/package/gluon-config-mode-outdoor/luasrc/lib/gluon/config-mode/wizard/0250-outdoor.lua b/package/gluon-config-mode-outdoor/luasrc/lib/gluon/config-mode/wizard/0250-outdoor.lua
index 947fa0b4ac..724b1798bf 100644
--- a/package/gluon-config-mode-outdoor/luasrc/lib/gluon/config-mode/wizard/0250-outdoor.lua
+++ b/package/gluon-config-mode-outdoor/luasrc/lib/gluon/config-mode/wizard/0250-outdoor.lua
@@ -2,7 +2,7 @@ return function(form, uci)
 	local platform = require 'gluon.platform'
 	local wireless = require 'gluon.wireless'
 
-	if not (platform.is_outdoor_device() and wireless.device_uses_11a(uci)) then
+	if not (platform.is_outdoor_device() and wireless.device_uses_band(uci, '5g')) then
 		-- only visible on wizard for outdoor devices
 		return
 	end
diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/180-outdoors b/package/gluon-core/luasrc/lib/gluon/upgrade/180-outdoors
index 59704bffc5..41386aa78c 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/180-outdoors
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/180-outdoors
@@ -23,7 +23,7 @@ if sysconfig.gluon_version then
 	outdoor = false
 elseif config == 'preset' then
 	-- enable outdoor mode through presets on new installs
-	outdoor = platform.is_outdoor_device() and wireless.device_uses_11a(uci)
+	outdoor = platform.is_outdoor_device() and wireless.device_uses_band(uci, '5g')
 else
 	-- enable/disable outdoor mode unconditionally on new installs
 	outdoor = config
diff --git a/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua b/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua
index 5845bca7a7..5058ef5e7e 100644
--- a/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua
+++ b/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua
@@ -122,11 +122,11 @@ function M.device_uses_wlan(uci)
 	return ret
 end
 
-function M.device_uses_11a(uci)
+function M.device_uses_band(uci, band)
 	local ret = false
 
 	uci:foreach('wireless', 'wifi-device', function(radio)
-		if radio.band == '5g' then
+		if radio.band == band then
 			ret = true
 			return false
 		end
diff --git a/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua b/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
index 3f0169418a..96548348d8 100644
--- a/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
+++ b/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
@@ -130,7 +130,7 @@ uci:foreach('wireless', 'wifi-device', function(config)
 end)
 
 
-if wireless.device_uses_11a(uci) and not wireless.preserve_channels(uci) then
+if wireless.device_uses_band(uci, '5g') and not wireless.preserve_channels(uci) then
 	local r = f:section(Section, translate("Outdoor Installation"), translate(
 		"Configuring the node for outdoor use tunes the 5 GHz radio to a frequency "
 		.. "and transmission power that conforms with the local regulatory requirements. "

From 85fa841298bfcaa4270a816fa35eb80b9d3c90be Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Wed, 20 Aug 2025 01:08:00 +0200
Subject: [PATCH 02/13] gluon-core: clean up band counting

using band2g instead of band24 allows to retrieve the actual band name from the key.
Add a helper function to increment the correct counter
---
 .../luasrc/lib/gluon/upgrade/200-wireless       | 17 +++++++++++------
 1 file changed, 11 insertions(+), 6 deletions(-)

diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
index a4ca20219b..7e8c051197 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
@@ -12,19 +12,24 @@ local uci = require('simple-uci').cursor()
 if not sysconfig.gluon_version then
 	uci:delete_all('wireless', 'wifi-iface')
 
+	local radio_band_count = {band_2g=0, band_5g=0}
+
+	local function count_band(band)
+		radio_band_count[band] = radio_band_count[band] + 1
+	end
+
 	-- First count all radios with a fixed frequency band.
 	-- This is needed to distribute devices which have radios
 	-- capable of operating in the 2.4 GHz and 5 GHz band need
 	-- to be distributed evenly.
-	local radio_band_count = {band24=0, band5=0}
 	wireless.foreach_radio(uci, function(radio)
 		local hwmodes = iwinfo.nl80211.hwmodelist(wireless.find_phy(radio))
 		if hwmodes.g and not (hwmodes.a or hwmodes.ac) then
 			-- 2.4 GHz
-			radio_band_count["band24"] = radio_band_count["band24"] + 1
+			count_band("band_2g")
 		elseif (hwmodes.a or hwmodes.ac) and not hwmodes.g then
 			-- 5 GHz
-			radio_band_count["band5"] = radio_band_count["band5"] + 1
+			count_band("band_5g")
 		end
 	end)
 
@@ -35,13 +40,13 @@ if not sysconfig.gluon_version then
 		local hwmodes = iwinfo.nl80211.hwmodelist(wireless.find_phy(radio))
 		if (hwmodes.a or hwmodes.ac) and hwmodes.g then
 			-- Dualband radio
-			if radio_band_count["band24"] <= radio_band_count["band5"] then
+			if radio_band_count["band_2g"] <= radio_band_count["band_5g"] then
 				-- Assign radio to 2.4GHz band
-				radio_band_count["band24"] = radio_band_count["band24"] + 1
+				count_band("band_2g")
 				uci:set('wireless', radio_name, 'band', '2g')
 			else
 				-- Assign radio to 5GHz band
-				radio_band_count["band5"] = radio_band_count["band5"] + 1
+				count_band("band_5g")
 				uci:set('wireless', radio_name, 'band', '5g')
 			end
 		end

From a0985305ba0cf741cb7972b633ace461ac28ea15 Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Wed, 20 Aug 2025 00:04:08 +0200
Subject: [PATCH 03/13] gluon-core: add initialization script of gluon wireless
 roles

migrate existing wireless configuration to the new role settings.
Existing roles are migrated to gluon.wireless section if the radio band is not configured yet.

This also checks to disable mesh if the device is an outdoor device.

Private wlan settings are migrated even if the iface is disabled.
---
 .../lib/gluon/upgrade/022-wireless-roles      | 109 ++++++++++++++++++
 1 file changed, 109 insertions(+)
 create mode 100755 package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles

diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles b/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles
new file mode 100755
index 0000000000..72c39d960b
--- /dev/null
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles
@@ -0,0 +1,109 @@
+#!/usr/bin/lua
+
+local site = require 'gluon.site'
+local uci = require('simple-uci').cursor()
+local util = require 'gluon.util'
+local wireless = require 'gluon.wireless'
+
+-- Create wifi roles if they do not exist yet
+local default_wifi_roles = {
+	band_2g = {},
+	band_5g = {},
+}
+
+-- enable client if it is not disabled
+if site.wifi24.ap() and not site.wifi24.ap.disabled(false) then
+	util.add_to_set(default_wifi_roles['band_2g'], 'client')
+end
+
+if site.wifi5.ap() and not site.wifi5.ap.disabled(false) then
+	util.add_to_set(default_wifi_roles['band_5g'], 'client')
+end
+
+-- enable mesh if it is not disabled
+if site.wifi24.mesh() and not site.wifi24.mesh.disabled(false) then
+	util.add_to_set(default_wifi_roles['band_2g'], 'mesh')
+end
+if site.wifi5.mesh() and not site.wifi5.mesh.disabled(false) then
+	util.add_to_set(default_wifi_roles['band_5g'], 'mesh')
+end
+
+local role_iface_names = {
+	client = 'client_radio',
+	mesh = 'mesh_radio',
+	private = 'wan_radio',
+}
+
+local function migrate_wifi_to_roles(band)
+	local band_roles = {}
+	local config_exists = false
+	wireless.foreach_radio(uci, function(radio)
+		if radio.band ~= band then
+			return
+		end
+		local radio_name = radio['.name']
+		local suffix = radio_name:match('^radio(%d+)$')
+
+		for role, iface_prefix in pairs(role_iface_names) do
+			local iface_name = iface_prefix .. suffix
+			if uci:get('wireless', iface_prefix .. suffix) then
+				config_exists = true
+				local role_disabled = uci:get('wireless', iface_prefix .. suffix, 'disabled') or 0
+				if role_disabled == '0' then
+					util.add_to_set(band_roles, role)
+					if role == 'private' then
+						local ssid = uci:get('wireless', iface_name, 'ssid')
+						local key = uci:get('wireless', iface_name, 'key')
+						local encryption = uci:get('wireless', iface_name, 'encryption')
+						local mfp = uci:get('wireless', iface_name, 'ieee80211w')
+						uci:set('gluon', 'wireless', 'private_ssid', ssid)
+						uci:set('gluon', 'wireless', 'private_key', key)
+						uci:set('gluon', 'wireless', 'private_encryption', encryption)
+						uci:set('gluon', 'wireless', 'private_mfp', mfp)
+					end
+				end
+			end
+		end
+	end)
+
+	-- if no config exists for this band, we apply the default
+	if not config_exists then
+		band_roles = default_wifi_roles["band_" .. band]
+	end
+
+	uci:section('gluon', 'wireless_band', "band_" .. band, {
+		role = band_roles,
+	})
+end
+
+for band_name, _ in pairs(default_wifi_roles) do
+	local raw_band = band_name:gsub("^band_", "")
+	if not uci:get('gluon', band_name) and wireless.device_uses_band(uci, raw_band) then
+		-- remove when upgrades are only possible from v2025.1 onwards
+		migrate_wifi_to_roles(raw_band)
+	end
+end
+
+
+local function is_outdoor()
+	return uci:get_bool('gluon', 'wireless', 'outdoor')
+end
+
+-- Fix invalid mesh role configurations for outdoor
+uci:foreach('gluon', 'wireless_band', function(band)
+	local function has_role(role)
+		if band.role == nil then
+			return false
+		end
+
+		return util.contains(band.role, role)
+	end
+
+	if band['.name'].name=='5g' and has_role('mesh') and is_outdoor() then
+		-- remove 'mesh' role
+		util.remove_from_set(band.role, 'mesh')
+		uci:set('gluon', band['.name'], 'role', band.role)
+	end
+end)
+
+uci:save('gluon')

From 696c9311440c7342afed94c19c7661699deacb08 Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Sun, 21 Sep 2025 23:12:23 +0200
Subject: [PATCH 04/13] gluon-core: create wireless config from gluon wireless
 roles section

Delete all wifi-ifaces on reconfigure.
Enable mesh based on the configured wireless role.
---
 .../luasrc/lib/gluon/upgrade/200-wireless     | 133 +++++++-----------
 1 file changed, 54 insertions(+), 79 deletions(-)

diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
index 7e8c051197..2f13c743c7 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
@@ -3,55 +3,51 @@
 local util = require 'gluon.util'
 local wireless = require 'gluon.wireless'
 local site = require 'gluon.site'
-local sysconfig = require 'gluon.sysconfig'
 local iwinfo = require 'iwinfo'
 
 local uci = require('simple-uci').cursor()
 
--- Initial
-if not sysconfig.gluon_version then
-	uci:delete_all('wireless', 'wifi-iface')
+uci:delete_all('wireless', 'wifi-iface')
 
-	local radio_band_count = {band_2g=0, band_5g=0}
+local radio_band_count = {band_2g=0, band_5g=0}
 
-	local function count_band(band)
-		radio_band_count[band] = radio_band_count[band] + 1
+local function count_band(band)
+	radio_band_count[band] = radio_band_count[band] + 1
+end
+
+-- First count all radios with a fixed frequency band.
+-- This is needed to distribute devices which have radios
+-- capable of operating in the 2.4 GHz and 5 GHz band need
+-- to be distributed evenly.
+wireless.foreach_radio(uci, function(radio)
+	local hwmodes = iwinfo.nl80211.hwmodelist(wireless.find_phy(radio))
+	if hwmodes.g and not (hwmodes.a or hwmodes.ac) then
+		-- 2.4 GHz
+		count_band("band_2g")
+	elseif (hwmodes.a or hwmodes.ac) and not hwmodes.g then
+		-- 5 GHz
+		count_band("band_5g")
 	end
+end)
 
-	-- First count all radios with a fixed frequency band.
-	-- This is needed to distribute devices which have radios
-	-- capable of operating in the 2.4 GHz and 5 GHz band need
-	-- to be distributed evenly.
-	wireless.foreach_radio(uci, function(radio)
-		local hwmodes = iwinfo.nl80211.hwmodelist(wireless.find_phy(radio))
-		if hwmodes.g and not (hwmodes.a or hwmodes.ac) then
-			-- 2.4 GHz
+-- Use the number of all fixed 2.4G GHz and 5 GHz radios to
+-- distribute dualband radios in this step.
+wireless.foreach_radio(uci, function(radio)
+	local radio_name = radio['.name']
+	local hwmodes = iwinfo.nl80211.hwmodelist(wireless.find_phy(radio))
+	if (hwmodes.a or hwmodes.ac) and hwmodes.g then
+		-- Dualband radio
+		if radio_band_count["band_2g"] <= radio_band_count["band_5g"] then
+			-- Assign radio to 2.4GHz band
 			count_band("band_2g")
-		elseif (hwmodes.a or hwmodes.ac) and not hwmodes.g then
-			-- 5 GHz
+			uci:set('wireless', radio_name, 'band', '2g')
+		else
+			-- Assign radio to 5GHz band
 			count_band("band_5g")
+			uci:set('wireless', radio_name, 'band', '5g')
 		end
-	end)
-
-	-- Use the number of all fixed 2.4G GHz and 5 GHz radios to
-	-- distribute dualband radios in this step.
-	wireless.foreach_radio(uci, function(radio)
-		local radio_name = radio['.name']
-		local hwmodes = iwinfo.nl80211.hwmodelist(wireless.find_phy(radio))
-		if (hwmodes.a or hwmodes.ac) and hwmodes.g then
-			-- Dualband radio
-			if radio_band_count["band_2g"] <= radio_band_count["band_5g"] then
-				-- Assign radio to 2.4GHz band
-				count_band("band_2g")
-				uci:set('wireless', radio_name, 'band', '2g')
-			else
-				-- Assign radio to 5GHz band
-				count_band("band_5g")
-				uci:set('wireless', radio_name, 'band', '5g')
-			end
-		end
-	end)
-end
+	end
+end)
 
 local function is_outdoor()
 	return uci:get_bool('gluon', 'wireless', 'outdoor')
@@ -90,31 +86,13 @@ local function get_htmode(radio)
 	return 'HT20'
 end
 
-local function is_disabled(name)
-	if not uci:get('wireless', name) then
-		return nil
-	end
-
-	return uci:get_bool('wireless', name, 'disabled')
-end
-
--- Returns the first argument that is not nil; don't call without any non-nil arguments!
-local function first_non_nil(first, ...)
-	if first ~= nil then
-		return first
-	end
-
-	return first_non_nil(...)
-end
-
-
 local function delete_ibss(radio_name)
 	local name = 'ibss_' .. radio_name
 
 	uci:delete('wireless', name)
 end
 
-local function configure_mesh(config, radio, index, suffix, disabled)
+local function configure_mesh(config, radio, index, suffix)
 	local radio_name = radio['.name']
 	local name = 'mesh_' .. radio_name
 
@@ -147,7 +125,7 @@ local function configure_mesh(config, radio, index, suffix, disabled)
 		basic_rate = { config.mcast_rate },
 		mcast_rate = config.mcast_rate,
 		ifname = suffix and 'mesh' .. suffix,
-		disabled = disabled,
+		disabled = false,
 		macfilter = macfilter,
 		maclist = maclist,
 	})
@@ -169,17 +147,11 @@ local function fixup_wan(radio, index)
 	uci:set('wireless', name, 'macaddr', macaddr)
 end
 
-local function configure_mesh_wireless(radio, index, config, disabled)
+local function configure_mesh_wireless(radio, index, config)
 	local radio_name = radio['.name']
 	local suffix = radio_name:match('^radio(%d+)$')
 
-	configure_mesh(config.mesh(), radio, index, suffix,
-		first_non_nil(
-			disabled,
-			is_disabled('mesh_' .. radio_name),
-			config.mesh.disabled(false)
-		)
-	)
+	configure_mesh(config.mesh(), radio, index, suffix)
 end
 
 local function set_channels(radio, radio_name, config)
@@ -225,27 +197,30 @@ wireless.foreach_radio(uci, function(radio, index, config)
 	uci:delete('wireless', radio_name, 'basic_rate')
 
 	local band = radio.band
+	-- now we assign radios to roles
+	local roles = uci:get('gluon', 'band_' .. band, 'role')
+
 	if band == '2g' then
 		uci:set('wireless', radio_name, 'legacy_rates', false)
-		configure_mesh_wireless(radio, index, config)
-	elseif (band == '5g') then
-		-- ToDo: Remove in v2024.x
-		local hostapd_options = uci:get_list('wireless', radio_name, 'hostapd_options')
-		util.remove_from_set(hostapd_options, 'country3=0x4f')
-		uci:set_list('wireless', radio_name, 'hostapd_options', hostapd_options)
-
-		if is_outdoor() then
-			-- enforce outdoor channels by filtering the regdom for outdoor channels
-			uci:set('wireless', radio_name, 'country3', '0x4f')
-			configure_mesh_wireless(radio, index, config, true)
-		else
-			uci:delete('wireless', radio_name, 'country3')
+	end
+
+	if roles and util.contains(roles, 'mesh') then
+		if band == '2g' then
 			configure_mesh_wireless(radio, index, config)
+		elseif (band == '5g') then
+			if is_outdoor() then
+				-- enforce outdoor channels by filtering the regdom for outdoor channels
+				uci:set('wireless', radio_name, 'country3', '0x4f')
+			else
+				uci:delete('wireless', radio_name, 'country3')
+				configure_mesh_wireless(radio, index, config)
+			end
 		end
 	end
 
 	uci:set('wireless', radio_name, 'beacon_int', beacon_interval)
 
+	uci:set('wireless', "wan_" .. radio_name, 'disabled', util.contains(roles, 'private'))
 	fixup_wan(radio, index)
 end)
 

From 5313d5de9204ef9fc57c9937739bb16948ba3335 Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Sun, 21 Sep 2025 23:13:28 +0200
Subject: [PATCH 05/13] gluon-core: implement detection of client_only_radio

This adds support for the configuration of the third radio on triband routers.
The client_only_radio is extended from the existing draft to theoretically also support 2g multi-band.

If there is an additional radio which does not support the mesh channel, we can configure it on a
different channel. In this case, the other radio on the same band supporting the channel turns off its
client radio and is mesh only.
---
 .../upgrade/320-gluon-client-bridge-wireless  | 21 +++++++++++++++++++
 .../luasrc/lib/gluon/upgrade/200-wireless     |  2 +-
 .../luasrc/usr/lib/lua/gluon/wireless.lua     | 10 +++++++++
 3 files changed, 32 insertions(+), 1 deletion(-)

diff --git a/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless b/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless
index bc5a7e249e..9aef5a1ac0 100755
--- a/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless
+++ b/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless
@@ -1,9 +1,21 @@
 #!/usr/bin/lua
 
 local wireless = require 'gluon.wireless'
+local util = require 'gluon.util'
 
 local uci = require('simple-uci').cursor()
 
+local client_only_radio = {band_2g=0, band_5g=0}
+
+local function count_band(band)
+	client_only_radio[band] = client_only_radio[band] + 1
+end
+
+wireless.foreach_radio(uci, function(radio, _, config)
+	if not wireless.supports_channel(radio, config.channel()) then
+		count_band("band_" .. radio.band)
+	end
+end)
 
 local function is_disabled(config, name)
 	if uci:get('wireless', name) then
@@ -105,6 +117,15 @@ end
 
 wireless.foreach_radio(uci, function(radio, index, config)
 	local radio_name = radio['.name']
+	-- now we assign radios to roles
+	local band = radio.band
+	local roles = uci:get('gluon', band, 'role')
+
+	-- if there is a radio which does not support the mesh channel, we can configure it on a different channel
+	-- in this case, the other radio on the same band supporting the channel turns off its client radio and is mesh only
+	local disable_client = wireless.supports_channel(radio, config.channel()) and client_only_radio['band' .. band] > 0
+
+	uci:set('wireless', 'client_' .. radio_name, 'disabled', util.contains(roles, 'client') and not disable_client)
 
 	configure_ap(radio, index, config, radio_name)
 
diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
index 2f13c743c7..32d07354e7 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
@@ -208,7 +208,7 @@ wireless.foreach_radio(uci, function(radio, index, config)
 		if band == '2g' then
 			configure_mesh_wireless(radio, index, config)
 		elseif (band == '5g') then
-			if is_outdoor() then
+			if is_outdoor() or not wireless.supports_channel(radio, config.channel()) then
 				-- enforce outdoor channels by filtering the regdom for outdoor channels
 				uci:set('wireless', radio_name, 'country3', '0x4f')
 			else
diff --git a/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua b/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua
index 5058ef5e7e..0575203371 100644
--- a/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua
+++ b/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua
@@ -39,6 +39,16 @@ local function get_wlan_mac_from_driver(radio, vif)
 	return addresses[vif+1]
 end
 
+function M.supports_channel(radio, channel)
+	local phy = M.find_phy(radio)
+	for _, chan in ipairs(iwinfo.nl80211.freqlist(phy)) do
+		if channel == chan.channel then
+			return true
+		end
+	end
+	return false
+end
+
 local radio_mac_offsets = {
 	client = 0,
 	mesh = 1,

From bee385b922435221a0fd36d757373916ae0f423c Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Mon, 18 Aug 2025 09:29:31 +0200
Subject: [PATCH 06/13] gluon-core: add hop_penalty support for mesh interfaces
 and fix preserve_channels

allow setting a batadv_hop_penalty for the gluon_mesh wireless interfaces as well
---
 package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
index 32d07354e7..8b66203aaa 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
@@ -110,8 +110,11 @@ local function configure_mesh(config, radio, index, suffix)
 		return
 	end
 
+	local batadv_hop_penalty = uci:get('gluon', 'band_' .. radio.band, 'batadv_hop_penalty') or 0
+
 	uci:section('network', 'interface', name, {
 		proto = 'gluon_mesh',
+		hop_penalty = batadv_hop_penalty,
 	})
 
 	uci:section('wireless', 'wifi-iface', name, {

From b0d9910d403fae037d5ab999c42bae54a9652d9a Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Mon, 18 Aug 2025 11:34:32 +0200
Subject: [PATCH 07/13] gluon-client-bridge: do not create wireless config if
 client role is disabled

This simplifies the running config by only providing what is needed.
We throw away the old config on each reconfigure now anyway,
so we do not need to generate all disabled ifaces anymore.
---
 .../upgrade/320-gluon-client-bridge-wireless  | 40 ++++++++-----------
 1 file changed, 16 insertions(+), 24 deletions(-)

diff --git a/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless b/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless
index 9aef5a1ac0..6c2df3fb89 100755
--- a/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless
+++ b/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless
@@ -17,12 +17,15 @@ wireless.foreach_radio(uci, function(radio, _, config)
 	end
 end)
 
-local function is_disabled(config, name)
-	if uci:get('wireless', name) then
-		return uci:get_bool('wireless', name, 'disabled')
-	end
-
-	return config.disabled(false)
+local function is_disabled(radio, config)
+	-- now we assign radios to roles
+	local band = radio.band
+	local roles = uci:get('gluon', 'band_' .. band, 'role')
+	-- if there is a radio which does not support the mesh channel, we can configure it on a different channel
+	-- in this case, the other radio on the same band supporting the channel turns off its client radio and is mesh only
+	local disable_client = wireless.supports_channel(radio, config.channel()) and client_only_radio['band' .. band] > 0
+	local client_enabled = roles and util.contains(roles, 'client') and not disable_client
+	return config.ap.disabled(not client_enabled)
 end
 
 local function configure_ap(radio, index, config, radio_name)
@@ -30,7 +33,6 @@ local function configure_ap(radio, index, config, radio_name)
 	local suffix = radio_name:match('^radio(%d+)$')
 
 	local ap = config.ap
-	local disabled = is_disabled(ap, name)
 
 	uci:delete('wireless', name)
 
@@ -47,7 +49,7 @@ local function configure_ap(radio, index, config, radio_name)
 		ssid = ap.ssid(),
 		macaddr = macaddr,
 		ifname = suffix and 'client' .. suffix,
-		disabled = disabled or false,
+		disabled = false,
 	})
 end
 
@@ -57,8 +59,6 @@ local function configure_owe(radio, index, config, radio_name)
 
 	local ap = config.ap
 
-	local disabled = is_disabled(ap, 'client_' .. radio_name)
-
 	uci:delete('wireless', name)
 
 	-- Don't configure OWE in case our device
@@ -80,7 +80,7 @@ local function configure_owe(radio, index, config, radio_name)
 		ssid = ap.owe_ssid(),
 		macaddr = macaddr,
 		ifname = suffix and 'owe' .. suffix,
-		disabled = disabled or false,
+		disabled = false,
 		encryption = 'owe',
 		ieee80211w = 2,
 	})
@@ -117,20 +117,12 @@ end
 
 wireless.foreach_radio(uci, function(radio, index, config)
 	local radio_name = radio['.name']
-	-- now we assign radios to roles
-	local band = radio.band
-	local roles = uci:get('gluon', band, 'role')
-
-	-- if there is a radio which does not support the mesh channel, we can configure it on a different channel
-	-- in this case, the other radio on the same band supporting the channel turns off its client radio and is mesh only
-	local disable_client = wireless.supports_channel(radio, config.channel()) and client_only_radio['band' .. band] > 0
-
-	uci:set('wireless', 'client_' .. radio_name, 'disabled', util.contains(roles, 'client') and not disable_client)
+	if not is_disabled(radio, config) then
+		configure_ap(radio, index, config, radio_name)
 
-	configure_ap(radio, index, config, radio_name)
-
-	configure_owe(radio, index, config, radio_name)
-	configure_owe_transition_mode(config, radio_name)
+		configure_owe(radio, index, config, radio_name)
+		configure_owe_transition_mode(config, radio_name)
+	end
 end)
 
 uci:save('wireless')

From 84567721a3b8d8690042667fcb9b3ab4ad4e5b84 Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Sat, 16 Aug 2025 18:14:22 +0200
Subject: [PATCH 08/13] gluon-private-wifi: add uci wireless config to
 gluon-private-wifi

Recreate private wifi config using upgrade scripts.
This reduces the coupling of gluon-web-private-wifi, which only sets the role of private wifi now.
---
 docs/features/private-wlan.rst                | 24 ++++----
 .../luasrc/lib/gluon/upgrade/200-wireless     | 19 ------
 .../lib/gluon/upgrade/325-gluon-private-wifi  | 49 +++++++++++++---
 .../config-mode/model/admin/privatewifi.lua   | 58 ++++++++-----------
 4 files changed, 75 insertions(+), 75 deletions(-)

diff --git a/docs/features/private-wlan.rst b/docs/features/private-wlan.rst
index d11c9c1cd8..f0f2ad9094 100644
--- a/docs/features/private-wlan.rst
+++ b/docs/features/private-wlan.rst
@@ -15,27 +15,23 @@ made optional or completely disabled in the advanced settings tab.
 The private WLAN can be enabled through the config mode if the package ``gluon-web-private-wifi`` is installed.
 You may also enable a private WLAN using the command line::
 
-  RID=0
   SSID="privateWLANname"
   KEY="yoursecret1337password"
 
-  uci set wireless.wan_radio$RID=wifi-iface
-  uci set wireless.wan_radio$RID.device=radio$RID
-  uci set wireless.wan_radio$RID.network=wan
-  uci set wireless.wan_radio$RID.mode=ap
-  uci set wireless.wan_radio$RID.encryption=psk2
-  uci set wireless.wan_radio$RID.ssid="$SSID"
-  uci set wireless.wan_radio$RID.key="$KEY"
-  uci set wireless.wan_radio$RID.disabled=0
-  uci set wireless.wan_radio$RID.macaddr=$(lua -e "print(require('gluon.wireless').get_wlan_mac('wan_radio', $RID))")
-  uci commit
+  uci add_list gluon.band_2g.role=private
+  uci add_list gluon.band_5g.role=private
+  uci set gluon.wireless.private_encryption=psk2
+  uci set gluon.wireless.private_ssid="$SSID"
+  uci set gluon.wireless.private_key="$KEY"
+  uci commit gluon
+  gluon-reconfigure
   wifi
 
 Please replace ``$SSID`` by the name of the WLAN and ``$KEY`` by your passphrase (8-63 characters).
-If you have two radios (e.g. 2.4 and 5 GHz) you need to do this for radio0 and radio1.
+If you have two radios (e.g. 2.4 and 5 GHz) you need to add the role to both bands.
 
 It may also be disabled by running::
 
-  uci set wireless.wan_radio0.disabled=1
-  uci commit
+  uci del_list set gluon.band_2g.role=private
+  uci commit gluon
   wifi
diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
index 8b66203aaa..2b10ab4332 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
@@ -134,22 +134,6 @@ local function configure_mesh(config, radio, index, suffix)
 	})
 end
 
-local function fixup_wan(radio, index)
-	local radio_name = radio['.name']
-	local name = 'wan_' .. radio_name
-
-	if not uci:get('wireless', name) then
-		return
-	end
-
-	local macaddr = wireless.get_wlan_mac('wan_radio', index, radio)
-	if not macaddr then
-		return
-	end
-
-	uci:set('wireless', name, 'macaddr', macaddr)
-end
-
 local function configure_mesh_wireless(radio, index, config)
 	local radio_name = radio['.name']
 	local suffix = radio_name:match('^radio(%d+)$')
@@ -222,9 +206,6 @@ wireless.foreach_radio(uci, function(radio, index, config)
 	end
 
 	uci:set('wireless', radio_name, 'beacon_int', beacon_interval)
-
-	uci:set('wireless', "wan_" .. radio_name, 'disabled', util.contains(roles, 'private'))
-	fixup_wan(radio, index)
 end)
 
 
diff --git a/package/gluon-private-wifi/luasrc/lib/gluon/upgrade/325-gluon-private-wifi b/package/gluon-private-wifi/luasrc/lib/gluon/upgrade/325-gluon-private-wifi
index 2485847cd2..516a507cc9 100755
--- a/package/gluon-private-wifi/luasrc/lib/gluon/upgrade/325-gluon-private-wifi
+++ b/package/gluon-private-wifi/luasrc/lib/gluon/upgrade/325-gluon-private-wifi
@@ -2,22 +2,55 @@
 
 local uci = require("simple-uci").cursor()
 local wireless = require 'gluon.wireless'
+local util = require 'gluon.util'
 
-wireless.foreach_radio(uci, function(radio)
+local function configure_private_wifi(radio, index)
 	local radio_name = radio['.name']
 	local suffix = radio_name:match('^radio(%d+)$')
 	local name   = "wan_" .. radio_name
 
-	if not uci:get('wireless', name, 'device') then
-		return
+	local encryption = uci:get('gluon', 'wireless', 'private_encryption')
+	local ssid = uci:get('gluon', 'wireless', 'private_ssid')
+	local key = uci:get('gluon', 'wireless', 'private_key')
+	local mfp = wireless.device_supports_mfp(uci) and uci:get('gluon', 'wireless', 'private_mfp')
+
+	-- migrate encryption from Gluon v2023.2.x or older
+	-- remove this in 2027 or on first release supporting only upgrades from >=v2025.1.x
+	encryption = encryption and encryption:gsub("psk3", "sae")
+
+	local macaddr = wireless.get_wlan_mac('wan_radio', index, radio)
+	uci:section('wireless', 'wifi-iface', name, {
+		device     = radio_name,
+		network    = 'wan',
+		mode       = 'ap',
+		encryption = encryption,
+		ssid       = ssid,
+		key        = key,
+		macaddr    = macaddr,
+		ifname     = suffix and 'wl-wan' .. suffix,
+		disabled   = false,
+	})
+
+	-- hostapd-mini won't start in case 802.11w is configured
+	if mfp then
+		uci:set('wireless', name, 'ieee80211w', mfp)
+	else
+		uci:delete('wireless', name, 'ieee80211w')
 	end
+end
+
+wireless.foreach_radio(uci, function(radio, index)
+	local radio_name = radio['.name']
+	local name   = "wan_" .. radio_name
 
-	uci:set('wireless', name, 'ifname', suffix and 'wl-wan' .. suffix)
+	local roles = uci:get('gluon', 'band_' .. radio.band, 'role')
 
-	-- migrate encryption from Gluon v2023.2.x or older
-	-- remove in 2027 or on first release supporting only upgrades from >=v2025.1.x
-	local encryption = uci:get('wireless', name, 'encryption')
-	uci:set('wireless', name, 'encryption', encryption:gsub("psk3", "sae"))
+	if not roles or not util.contains(roles, 'private') then
+		uci:delete('wireless', name)
+		return
+	end
+
+	configure_private_wifi(radio, index)
 end)
 
 uci:save('wireless')
diff --git a/package/gluon-web-private-wifi/luasrc/lib/gluon/config-mode/model/admin/privatewifi.lua b/package/gluon-web-private-wifi/luasrc/lib/gluon/config-mode/model/admin/privatewifi.lua
index ba649fccee..4b031e80f1 100644
--- a/package/gluon-web-private-wifi/luasrc/lib/gluon/config-mode/model/admin/privatewifi.lua
+++ b/package/gluon-web-private-wifi/luasrc/lib/gluon/config-mode/model/admin/privatewifi.lua
@@ -38,12 +38,12 @@ warning:depends(enabled, true)
 local ssid = s:option(Value, "ssid", translate("Name (SSID)"))
 ssid:depends(enabled, true)
 ssid.datatype = "maxlength(32)"
-ssid.default = uci:get('wireless', primary_iface, "ssid")
+ssid.default = uci:get('gluon', 'wireless', 'private_ssid')
 
 local key = s:option(Value, "key", translate("Key"), translate("8-63 characters"))
 key:depends(enabled, true)
 key.datatype = "wpakey"
-key.default = uci:get('wireless', primary_iface, "key")
+key.default = uci:get('gluon', 'wireless', 'private_key')
 
 local encryption = s:option(ListValue, "encryption", translate("Encryption"))
 encryption:depends(enabled, true)
@@ -52,7 +52,7 @@ if wireless.device_supports_wpa3() then
 	encryption:value("sae-mixed", translate("WPA2 / WPA3"))
 	encryption:value("sae", translate("WPA3"))
 end
-encryption.default = uci:get('wireless', primary_iface, 'encryption') or "psk2"
+encryption.default = uci:get('gluon', 'wireless', 'private_encryption') or "psk2"
 
 local mfp = s:option(ListValue, "mfp", translate("Management Frame Protection"))
 mfp:depends(enabled, true)
@@ -61,41 +61,31 @@ if wireless.device_supports_mfp(uci) then
 	mfp:value("1", translate("Optional"))
 	mfp:value("2", translate("Required"))
 end
-mfp.default = uci:get('wireless', primary_iface, 'ieee80211w') or "0"
+mfp.default = uci:get('gluon', 'wireless', 'private_mfp') or "0"
 
 
 function f:write()
-	wireless.foreach_radio(uci, function(radio, index)
-		local radio_name = radio['.name']
-		local suffix = radio_name:match('^radio(%d+)$')
-		local name   = "wan_" .. radio_name
-
-		if enabled.data then
-			local macaddr = wireless.get_wlan_mac('wan_radio', index, radio)
-
-			uci:section('wireless', 'wifi-iface', name, {
-				device     = radio_name,
-				network    = 'wan',
-				mode       = 'ap',
-				encryption = encryption.data,
-				ssid       = ssid.data,
-				key        = key.data,
-				macaddr    = macaddr,
-				ifname     = suffix and 'wl-wan' .. suffix,
-				disabled   = false,
-			})
-
-			-- hostapd-mini won't start in case 802.11w is configured
-			if wireless.device_supports_mfp(uci) then
-				uci:set('wireless', name, 'ieee80211w', mfp.data)
-			else
-				uci:delete('wireless', name, 'ieee80211w')
-			end
-		else
-			uci:set('wireless', name, "disabled", true)
-		end
-	end)
+	local roles_2g = uci:get_list('gluon', 'band_2g', 'role')
+	local roles_5g = uci:get_list('gluon', 'band_5g', 'role')
+	if enabled.data then
+		-- set new private wifi configuration
+		uci:set('gluon', 'wireless', 'private_ssid', ssid.data)
+		uci:set('gluon', 'wireless', 'private_key', key.data)
+		uci:set('gluon', 'wireless', 'private_encryption', encryption.data)
+		uci:set('gluon', 'wireless', 'private_mfp', mfp.data)
+
+		util.add_to_set(roles_2g, 'private')
+		util.add_to_set(roles_5g, 'private')
+	else
+		util.remove_from_set(roles_2g, 'private')
+		util.remove_from_set(roles_5g, 'private')
+	end
+	uci:set_list('gluon', 'band_2g', 'role', roles_2g)
+	uci:set_list('gluon', 'band_5g', 'role', roles_5g)
 
+	uci:commit('gluon')
+	os.execute('/usr/bin/gluon-reconfigure')
+	uci:commit('network')
 	uci:commit('wireless')
 end
 

From e53074fd217637eaee467e79a68048792f43004f Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Sun, 21 Sep 2025 23:40:48 +0200
Subject: [PATCH 09/13] gluon-web-wifi-config: refactor for gluon roles

Adjust configuration for roles.
Allow setting client and mesh per band instead of per radio.
Run gluon-reconfigure on save to take care of everything
---
 .../config-mode/model/admin/wifi-config.lua   | 82 +++++++++----------
 1 file changed, 38 insertions(+), 44 deletions(-)

diff --git a/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua b/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
index 96548348d8..6afe2d0559 100644
--- a/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
+++ b/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
@@ -2,7 +2,7 @@ local iwinfo = require 'iwinfo'
 local uci = require("simple-uci").cursor()
 local site = require 'gluon.site'
 local wireless = require 'gluon.wireless'
-
+local util = require 'gluon.util'
 
 local function txpower_list(phy)
 	local list = iwinfo.nl80211.txpwrlist(phy) or { }
@@ -38,15 +38,47 @@ f:section(Section, nil, translate(
 
 local mesh_vifs_5ghz = {}
 
+local function add_or_remove_role(roles, role, disabled)
+	if disabled == true then
+		util.remove_from_set(roles, role)
+	elseif disabled == false then
+		util.add_to_set(roles, role)
+	end
+end
+
+local function vif_option(section, role_name, band, msg)
+	local config_roles = uci:get('gluon', band, 'role') or {}
+	local o = section:option(Flag, band .. '_' .. role_name .. '_enabled', msg)
+	o.default = util.contains(config_roles, role_name)
+
+	function o:write(data)
+		-- this does race for multiple options without additional read before write in o:write
+		local roles = uci:get('gluon', band, 'role') or {}
+		add_or_remove_role(roles, role_name, not data)
+		uci:set('gluon', band, 'role', roles)
+	end
+
+	return o
+end
+
+local bands = {}
 
 uci:foreach('wireless', 'wifi-device', function(config)
 	local radio = config['.name']
 
 	local is_5ghz = false
 	local title
-	if config.band == '2g' then
+	local band = config.band
+
+	if bands and util.contains(bands, band) then
+		return
+	end
+
+	table.insert(bands, band)
+
+	if band == '2g' then
 		title = translate("2.4GHz WLAN")
-	elseif config.band == '5g' then
+	elseif band == '5g' then
 		is_5ghz = true
 		title = translate("5GHz WLAN")
 	else
@@ -55,47 +87,9 @@ uci:foreach('wireless', 'wifi-device', function(config)
 
 	local p = f:section(Section, title)
 
-	local function filter_existing_interfaces(interfaces)
-		local out = {}
-		for _, interface in ipairs(interfaces) do
-			if uci:get('wireless', interface .. '_' .. radio) then
-				table.insert(out, interface)
-			end
-		end
-		return out
-	end
-
-	local function has_active_interfaces(interfaces)
-		for _, interface in ipairs(interfaces) do
-			if not uci:get_bool('wireless', interface .. '_' .. radio, 'disabled') then
-				return true
-			end
-		end
-		return false
-	end
-
-	local function vif_option(name, interfaces, msg)
-		local existing_interfaces = filter_existing_interfaces(interfaces)
-
-		if #existing_interfaces == 0 then
-			return
-		end
-
-		local o = p:option(Flag, radio .. '_' .. name .. '_enabled', msg)
-		o.default = has_active_interfaces(existing_interfaces)
-
-		function o:write(data)
-			for _, interface in ipairs(existing_interfaces) do
-				uci:set('wireless', interface .. '_' .. radio, 'disabled', not data)
-			end
-		end
-
-		return o
-	end
-
-	vif_option('client', {'client', 'owe'}, translate('Enable client network (access point)'))
+	vif_option(p, 'client', band, translate('Enable client network (access point)'))
 
-	local mesh_vif = vif_option('mesh', {'mesh'}, translate("Enable mesh network (802.11s)"))
+	local mesh_vif = vif_option(p, 'mesh', band, translate("Enable mesh network (802.11s)"))
 	if is_5ghz then
 		table.insert(mesh_vifs_5ghz, mesh_vif)
 	end
@@ -186,7 +180,7 @@ end
 
 function f:write()
 	uci:commit('gluon')
-	os.execute('/lib/gluon/upgrade/200-wireless')
+	os.execute('/usr/bin/gluon-reconfigure')
 	uci:commit('network')
 	uci:commit('wireless')
 end

From b020744ec8a4744f1955cd910c03c5d6a8e0d770 Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Sun, 21 Sep 2025 18:22:42 +0200
Subject: [PATCH 10/13] gluon-core: remove delete_ibss function

the function is not required anymore as the whole wireless config is removed on each reconfigure anyway.
The old ibss config is therefore deleted as well.
---
 package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless | 8 --------
 1 file changed, 8 deletions(-)

diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
index 2b10ab4332..b2c5679f59 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
@@ -86,12 +86,6 @@ local function get_htmode(radio)
 	return 'HT20'
 end
 
-local function delete_ibss(radio_name)
-	local name = 'ibss_' .. radio_name
-
-	uci:delete('wireless', name)
-end
-
 local function configure_mesh(config, radio, index, suffix)
 	local radio_name = radio['.name']
 	local name = 'mesh_' .. radio_name
@@ -158,8 +152,6 @@ end
 wireless.foreach_radio(uci, function(radio, index, config)
 	local radio_name = radio['.name']
 
-	delete_ibss(radio_name)
-
 	if not config() then
 		uci:set('wireless', radio_name, 'disabled', true)
 		return

From a141708c103cd92853eee3126d15d9f4497984e0 Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Tue, 19 Aug 2025 23:18:51 +0200
Subject: [PATCH 11/13] gluon-core: migrate txpower setting and allow setting
 of htmode

add a txpower and htmode per radio section, which can be set using for example

uci set gluon.wireless.htmode_radio0="HE40"

Similarly, txpower_radio0 can be adjusted.
The migration of the existing txpower value takes already place in 022-wireless-roles
---
 .../lib/gluon/upgrade/022-wireless-roles      | 13 +++
 .../luasrc/lib/gluon/upgrade/200-wireless     | 18 +++-
 .../config-mode/model/admin/wifi-config.lua   | 91 +++++++++----------
 3 files changed, 67 insertions(+), 55 deletions(-)

diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles b/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles
index 72c39d960b..15d31ea388 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles
@@ -64,6 +64,19 @@ local function migrate_wifi_to_roles(band)
 				end
 			end
 		end
+
+		local txpower = uci:get('wireless', radio_name, 'txpower')
+		if txpower ~= nil then
+			uci:set('gluon', 'wireless', 'txpower_' .. radio['.name'], txpower)
+		end
+
+		if radio.band == '5g' then
+			local outdoor_htmode = uci:get('gluon', 'wireless', 'outdoor_' .. radio['.name'] .. '_htmode')
+			if outdoor_htmode ~= nil then
+				uci:set('gluon', 'wireless', 'htmode_' .. radio['.name'], outdoor_htmode)
+				uci:delete('gluon', 'wireless', 'outdoor_' .. radio['.name'] .. '_htmode')
+			end
+		end
 	end)
 
 	-- if no config exists for this band, we apply the default
diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
index b2c5679f59..736f3635b9 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
@@ -62,16 +62,19 @@ local function get_channel(radio, config)
 	return config.channel()
 end
 
+local function get_txpower(radio)
+	return uci:get('gluon', 'wireless', 'txpower_' .. radio['.name'])
+end
+
 local function get_htmode(radio)
 	if wireless.preserve_channels(uci) then
 		return radio.htmode
 	end
 
-	if radio.band == '5g' and is_outdoor() then
-		local outdoor_htmode = uci:get('gluon', 'wireless', 'outdoor_' .. radio['.name'] .. '_htmode')
-		if outdoor_htmode ~= nil then
-			return outdoor_htmode
-		end
+
+	local htmode = uci:get('gluon', 'wireless', 'htmode_' .. radio['.name'])
+	if htmode ~= nil then
+		return htmode
 	end
 
 	local phy = wireless.find_phy(radio)
@@ -162,6 +165,11 @@ wireless.foreach_radio(uci, function(radio, index, config)
 		return
 	end
 
+	local txpower = get_txpower(radio)
+	if txpower ~= nil then
+		uci:set('wireless', radio_name, 'txpower', txpower)
+	end
+
 	local htmode = get_htmode(radio)
 	local beacon_interval = config.beacon_interval()
 
diff --git a/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua b/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
index 6afe2d0559..1e60065df2 100644
--- a/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
+++ b/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
@@ -70,30 +70,32 @@ uci:foreach('wireless', 'wifi-device', function(config)
 	local title
 	local band = config.band
 
-	if bands and util.contains(bands, band) then
-		return
-	end
-
-	table.insert(bands, band)
-
-	if band == '2g' then
-		title = translate("2.4GHz WLAN")
-	elseif band == '5g' then
-		is_5ghz = true
-		title = translate("5GHz WLAN")
+	local p
+	-- store section to add txpower and htmode to
+	if bands and bands[band] then
+		p = bands[band]
 	else
-		return
-	end
+		if band == '2g' then
+			title = translate("2.4GHz WLAN")
+		elseif band == '5g' then
+			is_5ghz = true
+			title = translate("5GHz WLAN")
+		else
+			return
+		end
 
-	local p = f:section(Section, title)
+		p = f:section(Section, title)
 
-	vif_option(p, 'client', band, translate('Enable client network (access point)'))
+		vif_option(p, 'client', band, translate('Enable client network (access point)'))
 
-	local mesh_vif = vif_option(p, 'mesh', band, translate("Enable mesh network (802.11s)"))
-	if is_5ghz then
-		table.insert(mesh_vifs_5ghz, mesh_vif)
+		local mesh_vif = vif_option(p, 'mesh', band, translate("Enable mesh network (802.11s)"))
+		if is_5ghz then
+			table.insert(mesh_vifs_5ghz, mesh_vif)
+		end
+		bands[band] = p
 	end
 
+	-- txpowerlist
 	local phy = wireless.find_phy(config)
 	if not phy then
 		return
@@ -104,8 +106,8 @@ uci:foreach('wireless', 'wifi-device', function(config)
 		return
 	end
 
-	local tp = p:option(ListValue, radio .. '_txpower', translate("Transmission power"))
-	tp.default = uci:get('wireless', radio, 'txpower') or 'default'
+	local tp = p:option(ListValue, 'txpower_' .. radio, translate("Transmission power").. ' (' .. radio .. ')')
+	tp.default = uci:get('gluon', 'wireless', 'txpower_' .. radio) or 'default'
 
 	tp:value('default', translate("(default)"))
 
@@ -119,7 +121,25 @@ uci:foreach('wireless', 'wifi-device', function(config)
 		if data == 'default' then
 			data = nil
 		end
-		uci:set('wireless', radio, 'txpower', data)
+		uci:set('gluon', 'wireless', 'txpower_' .. radio, data)
+	end
+
+	-- htmode
+	local ht = p:option(ListValue, 'htmode_' .. radio, translate('HT Mode') .. ' (' .. radio .. ')')
+	ht.default = uci:get('gluon', 'wireless', 'htmode_' .. radio) or 'default'
+	ht:value('default', translate("(default)"))
+
+	for mode, available in pairs(iwinfo.nl80211.htmodelist(phy)) do
+		if available then
+			ht:value(mode, mode)
+		end
+	end
+
+	function ht:write(data)
+		if data == 'default' then
+			data = nil
+		end
+		uci:set('gluon', 'wireless', 'htmode_' .. radio, data)
 	end
 end)
 
@@ -146,35 +166,6 @@ if wireless.device_uses_band(uci, '5g') and not wireless.preserve_channels(uci)
 	function outdoor:write(data)
 		uci:set('gluon', 'wireless', 'outdoor', data)
 	end
-
-	uci:foreach('wireless', 'wifi-device', function(config)
-		local radio = config['.name']
-		local band = uci:get('wireless', radio, 'band')
-
-		if band ~= '5g' then
-			return
-		end
-
-		local phy = wireless.find_phy(config)
-
-		local ht = r:option(ListValue, 'outdoor_htmode', translate('HT Mode') .. ' (' .. radio .. ')')
-		ht:depends(outdoor, true)
-		ht.default = uci:get('gluon', 'wireless', 'outdoor_' .. radio .. '_htmode') or 'default'
-
-		ht:value('default', translate("(default)"))
-		for mode, available in pairs(iwinfo.nl80211.htmodelist(phy)) do
-			if available then
-				ht:value(mode, mode)
-			end
-		end
-
-		function ht:write(data)
-			if data == 'default' then
-				data = nil
-			end
-			uci:set('gluon', 'wireless', 'outdoor_' .. radio .. '_htmode', data)
-		end
-	end)
 end
 
 

From dfa0ea8026e7111b4b9e3f489436c2c5fc6b9407 Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Sat, 6 Sep 2025 02:51:52 +0200
Subject: [PATCH 12/13] add vero w6m  to supported devices

---
 docs/user/supported_devices.rst | 4 ++++
 targets/mediatek-filogic        | 7 +++++++
 2 files changed, 11 insertions(+)

diff --git a/docs/user/supported_devices.rst b/docs/user/supported_devices.rst
index 71d8550909..52d312c962 100644
--- a/docs/user/supported_devices.rst
+++ b/docs/user/supported_devices.rst
@@ -313,6 +313,10 @@ lantiq-xway
 mediatek-filogic
 ----------------
 
+* Acer
+
+  - Vero-W6M
+
 * ASUS
 
   - RT-AX52
diff --git a/targets/mediatek-filogic b/targets/mediatek-filogic
index 2a399c500b..818bb01114 100644
--- a/targets/mediatek-filogic
+++ b/targets/mediatek-filogic
@@ -1,6 +1,13 @@
 config('TARGET_ROOTFS_INITRAMFS', true)
 config('TARGET_INITRAMFS_COMPRESSION_LZMA', true)
 
+-- Acer
+
+device('acer-vero-w6m', 'acer_vero-w6m', {
+	factory = false,
+})
+
+
 -- ASUS
 
 device('asus-rt-ax52', 'asus_rt-ax52', {

From 3a754d29732bb07a5c5eb131a9dff21bf4d16ff2 Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Mon, 18 Aug 2025 10:50:37 +0200
Subject: [PATCH 13/13] gluon-core: add basic support for 6GHz wifi

---
 package/gluon-client-bridge/check_site.lua       |  2 +-
 .../upgrade/320-gluon-client-bridge-wireless     | 16 +++++++++++-----
 package/gluon-core/check_site.lua                |  8 +++++++-
 .../luasrc/lib/gluon/upgrade/022-wireless-roles  |  4 ++++
 .../luasrc/lib/gluon/upgrade/200-wireless        |  7 +++++--
 .../luasrc/usr/lib/lua/gluon/wireless.lua        |  2 ++
 .../lib/gluon/upgrade/205-wireless-mesh-sae      |  2 +-
 .../config-mode/model/admin/privatewifi.lua      |  5 +++++
 .../config-mode/model/admin/wifi-config.lua      |  2 ++
 9 files changed, 38 insertions(+), 10 deletions(-)

diff --git a/package/gluon-client-bridge/check_site.lua b/package/gluon-client-bridge/check_site.lua
index b077f87bb5..8cbf2ffca0 100644
--- a/package/gluon-client-bridge/check_site.lua
+++ b/package/gluon-client-bridge/check_site.lua
@@ -7,7 +7,7 @@ end
 need_string_match(in_domain({'next_node', 'ip6'}), '^[%x:]+$', false)
 
 
-for _, config in ipairs({'wifi24', 'wifi5'}) do
+for _, config in ipairs({'wifi24', 'wifi5', 'wifi6'}) do
 	if need_table({config, 'ap'}, nil, false) then
 		need_boolean({config, 'ap', 'disabled'}, false)
 		if need_boolean({config, 'ap', 'owe_transition_mode'}, false) then
diff --git a/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless b/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless
index 6c2df3fb89..2118f4f4a9 100755
--- a/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless
+++ b/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless
@@ -5,7 +5,7 @@ local util = require 'gluon.util'
 
 local uci = require('simple-uci').cursor()
 
-local client_only_radio = {band_2g=0, band_5g=0}
+local client_only_radio = {band_2g=0, band_5g=0, band_6g=0}
 
 local function count_band(band)
 	client_only_radio[band] = client_only_radio[band] + 1
@@ -23,7 +23,7 @@ local function is_disabled(radio, config)
 	local roles = uci:get('gluon', 'band_' .. band, 'role')
 	-- if there is a radio which does not support the mesh channel, we can configure it on a different channel
 	-- in this case, the other radio on the same band supporting the channel turns off its client radio and is mesh only
-	local disable_client = wireless.supports_channel(radio, config.channel()) and client_only_radio['band' .. band] > 0
+	local disable_client = wireless.supports_channel(radio, config.channel()) and client_only_radio['band_' .. band] > 0
 	local client_enabled = roles and util.contains(roles, 'client') and not disable_client
 	return config.ap.disabled(not client_enabled)
 end
@@ -117,11 +117,17 @@ end
 
 wireless.foreach_radio(uci, function(radio, index, config)
 	local radio_name = radio['.name']
-	if not is_disabled(radio, config) then
-		configure_ap(radio, index, config, radio_name)
+
+	if not is_disabled(radio, config)  then
+		-- 6ghz does not support unencrypted client ifaces
+		if radio.band ~= '6g' then
+			configure_ap(radio, index, config, radio_name)
+		end
 
 		configure_owe(radio, index, config, radio_name)
-		configure_owe_transition_mode(config, radio_name)
+		if radio.band ~= '6g' then
+			configure_owe_transition_mode(config, radio_name)
+		end
 	end
 end)
 
diff --git a/package/gluon-core/check_site.lua b/package/gluon-core/check_site.lua
index cf72f05fc8..c72dab3774 100644
--- a/package/gluon-core/check_site.lua
+++ b/package/gluon-core/check_site.lua
@@ -31,7 +31,7 @@ need_string_match(in_domain({'prefix6'}), '^[%x:]+/64$')
 need_string_array_match(in_domain({'extra_prefixes6'}), '^[%x:]+/%d+$', false)
 
 local supported_rates = {6000, 9000, 12000, 18000, 24000, 36000, 48000, 54000}
-for _, config in ipairs({'wifi24', 'wifi5'}) do
+for _, config in ipairs({'wifi24', 'wifi5', 'wifi6'}) do
 	if need_table({config}, nil, false) then
 		need_string(in_site({'regdom'})) -- regdom is only required when wifi24 or wifi5 is configured
 		need_number({config, 'beacon_interval'}, false)
@@ -48,6 +48,12 @@ for _, config in ipairs({'wifi24', 'wifi5'}) do
 			need_one_of({config, 'channel'}, channels)
 			need_chanlist({config, 'outdoor_chanlist'}, channels, false)
 			need_one_of({config, 'outdoors'}, {true, false, 'preset'}, false)
+		elseif config == 'wifi6' then
+			local channels = {
+				1, 5, 9, 13, 17, 21, 25, 29, 33, 37, 41, 45, 49, 53, 57, 61,
+				65, 69, 73, 77, 81, 85, 89, 93 }
+			need_one_of({config, 'channel'}, channels)
+			need_string_match(in_domain({config, 'ap', 'owe_ssid'}), '^' .. ('.?'):rep(32) .. '$', false)
 		end
 
 		obsolete({config, 'supported_rates'}, '802.11b rates are disabled by default.')
diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles b/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles
index 15d31ea388..8a22f3fdcf 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles
@@ -9,6 +9,7 @@ local wireless = require 'gluon.wireless'
 local default_wifi_roles = {
 	band_2g = {},
 	band_5g = {},
+	band_6g = {},
 }
 
 -- enable client if it is not disabled
@@ -27,6 +28,9 @@ end
 if site.wifi5.mesh() and not site.wifi5.mesh.disabled(false) then
 	util.add_to_set(default_wifi_roles['band_5g'], 'mesh')
 end
+if not site.wifi6.mesh.disabled(false) then
+	util.add_to_set(default_wifi_roles['band_6g'], 'mesh')
+end
 
 local role_iface_names = {
 	client = 'client_radio',
diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
index 736f3635b9..a9f9530f5e 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
@@ -9,7 +9,7 @@ local uci = require('simple-uci').cursor()
 
 uci:delete_all('wireless', 'wifi-iface')
 
-local radio_band_count = {band_2g=0, band_5g=0}
+local radio_band_count = {band_2g=0, band_5g=0, band_6g=0}
 
 local function count_band(band)
 	radio_band_count[band] = radio_band_count[band] + 1
@@ -27,6 +27,9 @@ wireless.foreach_radio(uci, function(radio)
 	elseif (hwmodes.a or hwmodes.ac) and not hwmodes.g then
 		-- 5 GHz
 		count_band("band_5g")
+	elseif hwmodes.ax and not hwmodes.ac then
+		-- 6 GHz
+		count_band("band_6g")
 	end
 end)
 
@@ -192,7 +195,7 @@ wireless.foreach_radio(uci, function(radio, index, config)
 	end
 
 	if roles and util.contains(roles, 'mesh') then
-		if band == '2g' then
+		if band == '2g' or band == '6g' then
 			configure_mesh_wireless(radio, index, config)
 		elseif (band == '5g') then
 			if is_outdoor() or not wireless.supports_channel(radio, config.channel()) then
diff --git a/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua b/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua
index 0575203371..2fe50f76aa 100644
--- a/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua
+++ b/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua
@@ -89,6 +89,8 @@ function M.foreach_radio(uci, f)
 			f(radio, index-1, site.wifi24)
 		elseif band == '5g' then
 			f(radio, index-1, site.wifi5)
+		elseif band == '6g' then
+			f(radio, index-1, site.wifi6)
 		end
 	end
 end
diff --git a/package/gluon-mesh-wireless-sae/luasrc/lib/gluon/upgrade/205-wireless-mesh-sae b/package/gluon-mesh-wireless-sae/luasrc/lib/gluon/upgrade/205-wireless-mesh-sae
index a229896694..99f69f3ca1 100755
--- a/package/gluon-mesh-wireless-sae/luasrc/lib/gluon/upgrade/205-wireless-mesh-sae
+++ b/package/gluon-mesh-wireless-sae/luasrc/lib/gluon/upgrade/205-wireless-mesh-sae
@@ -20,7 +20,7 @@ wireless.foreach_radio(uci, function(radio)
 		uci:delete('wireless', vif, 'encryption')
 		uci:delete('wireless', vif, 'key')
 
-		if enable then
+		if enable or radio.band == '6g' then
 			configure_sae(vif)
 		end
 	end
diff --git a/package/gluon-web-private-wifi/luasrc/lib/gluon/config-mode/model/admin/privatewifi.lua b/package/gluon-web-private-wifi/luasrc/lib/gluon/config-mode/model/admin/privatewifi.lua
index 4b031e80f1..690d362d8c 100644
--- a/package/gluon-web-private-wifi/luasrc/lib/gluon/config-mode/model/admin/privatewifi.lua
+++ b/package/gluon-web-private-wifi/luasrc/lib/gluon/config-mode/model/admin/privatewifi.lua
@@ -67,8 +67,10 @@ mfp.default = uci:get('gluon', 'wireless', 'private_mfp') or "0"
 function f:write()
 	local roles_2g = uci:get_list('gluon', 'band_2g', 'role')
 	local roles_5g = uci:get_list('gluon', 'band_5g', 'role')
+	local roles_6g = uci:get_list('gluon', 'band_6g', 'role')
 	if enabled.data then
 		-- set new private wifi configuration
+		-- TODO should this be set everytime? To reset this with empty values?
 		uci:set('gluon', 'wireless', 'private_ssid', ssid.data)
 		uci:set('gluon', 'wireless', 'private_key', key.data)
 		uci:set('gluon', 'wireless', 'private_encryption', encryption.data)
@@ -76,12 +78,15 @@ function f:write()
 
 		util.add_to_set(roles_2g, 'private')
 		util.add_to_set(roles_5g, 'private')
+		util.add_to_set(roles_6g, 'private')
 	else
 		util.remove_from_set(roles_2g, 'private')
 		util.remove_from_set(roles_5g, 'private')
+		util.remove_from_set(roles_6g, 'private')
 	end
 	uci:set_list('gluon', 'band_2g', 'role', roles_2g)
 	uci:set_list('gluon', 'band_5g', 'role', roles_5g)
+	uci:set_list('gluon', 'band_6g', 'role', roles_6g)
 
 	uci:commit('gluon')
 	os.execute('/usr/bin/gluon-reconfigure')
diff --git a/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua b/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
index 1e60065df2..afa3d81e30 100644
--- a/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
+++ b/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
@@ -77,6 +77,8 @@ uci:foreach('wireless', 'wifi-device', function(config)
 	else
 		if band == '2g' then
 			title = translate("2.4GHz WLAN")
+		elseif band == '6g' then
+			title = translate("6GHz WLAN")
 		elseif band == '5g' then
 			is_5ghz = true
 			title = translate("5GHz WLAN")
